espresso/environment/4_confirmation_integrity_with_reorgs_test.go

```diff
@@ -0,0 +1,212 @@
+func BatchHash(b *derive.SingularBatch) string {
+
+	// Concatenate the transactions and other relevant metadata of the batch
+	str := ""
+
+	for _, tx := range b.Transactions {
+		str = str + tx.String()
+	}
+	str = str + b.EpochHash.String()
+	str = str + strconv.Itoa(int(b.Timestamp))
+	str = str + b.ParentHash.String()
+
+	h := sha256.New()
+	h.Write([]byte(str))
+	hash := h.Sum(nil)
+
+	res := hex.EncodeToString(hash)
+
+	return res
+}
+
+// This function computes a list where at least a few batches are sent to unfinalized L1 blocks.
+// It works as follows:
+//  1. Pick the current L1 block number. Store it so that later we can reorg back to this point
+//  2. Start from the latest L2 block height.
+//  3. Pick the blocks from this height the subsequent. These blocks are unsafe (i.e. not yet sent to L1)
+//  4. Do this until the "finality" window closes i.e. before the number of new L1 blocks is bigger than  L1FinalizedDistance
+//  5. While doing this also send transactions to the sequencer
+//  6. Return the list of batch hashes, the L1 block number to reorg to and also the L2 block height determined in step 2
+//     @param ctx,t,system standard parameters to execute a test
+//     @param l1Client L1 client used to fetch L1 block numbers
+//     @param l2Seq sequencer used to send transactions
+//     @param l2Verif OP node we monitor for unsafe blocks
+//     @return batches list of hashes of the batches
+//     @return l1HeightStart L1 block number collected in step 1.
+//     @return unsafeL2BlockNumber L2 block number collected in step 2.
+func collectBatchesPublishedOnUnfinalizedL1Blocks(ctx context.Context, t *testing.T, system *e2esys.System, l1Client *ethclient.Client, l2Seq *ethclient.Client, l2Verif *ethclient.Client) ([]string, uint64, uint64) {
+	var batches []string
+
+	l1Height, err := l1Client.BlockNumber(ctx)
+	require.NoError(t, err)
+	l1HeightStart := l1Height
+	log.Info("L1 height to reorg to", "height", l1HeightStart)
+	// Keep monitoring L2 blocks while L1 is producing unfinalized blocks
+	i := int64(0)
+	// Fetch height of the most recent block which is unsafe and which batch will be sent to L1 at a later stage
+	unsafeL2BlockNumber, err := l2Verif.BlockNumber(ctx)
+
+	nonce := uint64(0)
+	addressAlice := system.Cfg.Secrets.Addresses().Alice
+
+	require.NoError(t, err)
+	for (l1Height - l1HeightStart) < system.Cfg.L1FinalizedDistance {
+		height := uint64(i) + unsafeL2BlockNumber
+
+		//Send some transactions to fill the batches
+		receipt := helpers.SendL2TxWithID(t, system.Cfg.L2ChainIDBig(), l2Seq, system.Cfg.Secrets.Bob, func(opts *helpers.TxOpts) {
+			opts.Nonce = nonce
+			opts.ToAddr = &addressAlice
+			opts.Value = new(big.Int).SetUint64(1)
+		})
+		nonce++
+		log.Info("Receipt", "value", receipt)
+
+		l2Head, err := geth.WaitForBlockToBeSafe(new(big.Int).SetUint64(height), l2Verif, 10*time.Second)
+		require.NoError(t, err)
+		time.Sleep(500 * time.Millisecond)
```

Comment:
Does this `Sleep` correspond to any step in the function description? Or is it for delaying the batch insertion below or the next iteration?

Reply:
Good catch, this call to `time.Sleep` was part of earlier version of this test and is unnecessary.
Removed in fd47567d1.